ykrt/pass_finder/try_passes.py

```diff
@@ -66,53 +66,96 @@ def log(logf, s):
 def executable_exists(cmd):
     return shutil.which(cmd) is not None
 
-def get_all_passes(is_prelink):
-    cmd = get_opt_cmd(is_prelink) 
-    p = Popen(cmd, shell=True, stdout=PIPE, close_fds=True)
-    sout, serr = p.communicate() 
-    assert(p.returncode == 0)
+def split_passes(passes_string):
+    parts = []           
+    temp_part = []      
+    paren_stack = []     
+    angle_stack = []    
+
+    i = 0  
+    while i < len(passes_string):
+        char = passes_string[i]
+
+        if char == '(':
+            paren_stack.append(char)
+        elif char == ')':
+            if paren_stack:
+                paren_stack.pop()
+
+        if char == '<':
+            angle_stack.append(char)
+        elif char == '>':
+            if angle_stack:
+                angle_stack.pop()
+
+        if char == ',' and not paren_stack and not angle_stack:
+            part = ''.join(temp_part).strip()
+            if part:  
+                parts.append(part)
+            temp_part = []  
+        else: 
+            temp_part.append(char)
+
+        i += 1  
+    
+    part = ''.join(temp_part).strip()
+    if part:  
+        parts.append(part)
 
-    sout = sout.decode()
-    pass_descrs = [x.strip() for x in sout.strip().split("\n")]
+    return parts
 
+def get_all_passes(is_prelink):
+    #cmd = get_opt_cmd(is_prelink)
+    cmd = "opt -passes='lto-pre-link<O2>' -print-pipeline-passes < /dev/null 2>/dev/null"
+    sout = subprocess.run(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL)
+    sout = sout.stdout.decode('utf-8')
+    pass_descrs = split_passes(sout)
+    print(f"\033[91m{pass_descrs}\033[0m")
```

Comment:
What do the control codes do?

Reply:
the codes prints statements in ansi color, makes it easier to find some line in huge stdout dump (I'll remove this in final version).

Reply:
I've done this before too. I'm OK with it for this little script.